stdlib/universe/filter.go

```diff
@@ -345,15 +319,6 @@ func (t *filterTransformation) filter(cr flux.ColReader, record values.Object, i
 	return bitset, nil
 }
 
-func (t *filterTransformation) wait(ctx context.Context, done <-chan struct{}) error {
-	select {
-	case <-done:
-		return nil
-	case <-ctx.Done():
-		return ctx.Err()
-	}
-}
-
```

Comment:
I'm still not clear why waiting for a downstream transformation would cause a deadlock: 

> ...the dispatcher which, for simple queries with one root, would only
allocate a single goroutine for executing the query. This meant that
since the filter() call never yielded...

Was this for cases where there is no transformation after the call to `filter()`?

Reply:
Waiting for the downstream transformation ended up causing a problem because of the following flow.

1. During planning, the planner decided the concurrency of the query was 1. That meant only one transformation could run at a time. I didn't understand the code completely, but I think it decided it was 1 because there was only one root (the yield call). I don't know if the planner refers to the yield as the root or the from as the root, but there was only one of each of those so it set the concurrency to 1.
2. The transformation started. Filter processed a multi-buffer table. It told the downstream transformation to process a table and then waited for it to be processed.
3. Filter held the only open goroutine in the dispatcher so the next transformation never ran. That meant the table was never read and so filter never released the goroutine. I was aware this could happen, but I misread the code and thought the concurrency was always set to 10. There's a number called throughput that is set to 10, but it has nothing to do with the number of goroutines.
4. `filter |> yield` was not affected because yield isn't a real transformation. It creates a result that is returned to the user. That code is not controlled by the dispatcher so it didn't matter that there was no open goroutine.

The changes that were made were:

1. Properly fix the filter function so it is thread safe.
2. Filter could go back to the old code that did not wait for the table to finish being processed. This frees up the goroutine again so that the next transformation can be run.
3. Add `debug.pass()` to the test. This forces an additional transformation that does nothing. I used `debug.pass()` to debug the push down filter and found the same thing useful for this test to ensure that the downstream transformation wasn't blocked by filter. Since `debug.pass()` is a real transformation, it takes up space in the dispatcher and ensures that any other transformation would be able to run too.